sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/MapElements.java

```diff
@@ -162,4 +171,89 @@ public void populateDisplayData(DisplayData.Builder builder) {
       builder.include("fn", (HasDisplayData) originalFnForDisplayData);
     }
   }
+
+  /**
+   * Sets a {@link TupleTag} to associate with successes, converting this {@link PTransform} into
+   * one that returns a {@link PCollectionTuple}. This allows you to make subsequent {@link
+   * WithFailures#withFailureTag(TupleTag, Class, Class[])} calls to capture thrown exceptions to
+   * failure collections.
+   */
+  public WithFailures withSuccessTag(TupleTag<OutputT> successTag) {
+    return new WithFailures(successTag, TaggedExceptionsList.empty());
+  }
```

Comment:
consider adding both withSuccessTag and withFailureTag at the top level so that users can call them in either order.

Reply:
I had wanted to avoid that so that we could verify state at compile time. Calling `withFailureTag` without first setting the success tag would require us to check state at runtime and fail. 

This situation seems similar to other places in the Java SDK where we use intermediate objects to constrain possible misconfigurations at compile time. I'm thinking about `FileIO.writeDynamic` as one example where methods must be called in a certain order and this allows compile-time checking.

Reply:
Calling `withFailureTag` without first setting the success tag doesn't have to be an invalid state. Someone might have code that always throws an exception, which could return a PCollectionTuple of only `Failure`s.

Reply:
Fair point. I can add `withFailureTag` to the main class. There's getting to be quite a lot of duplicate code between MapElements, FlatMapElements, and Filter. I'm wondering whether we've reached the point of refactoring to have all three depend on a common abstract `SingleMessageTransform` base class that could include all the machinery for generating `WithFailures` instances.

I'm going to post to the mailing list about that.